File: August2020/ReorderList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

// Time Complexity is O(n)
class ReorderList {
    public void reorderList(ListNode head) {

        if(head ==null || head.next ==null || head.next.next ==null){
            return ;
        }
        ListNode fast = head; // Step 1 => Break the list into 2 parts using fast and slow pointer techinque
        ListNode slow = head;

        ListNode slowPrev =null;
        while(fast!=null && fast.next !=null){
            slowPrev = slow;
            slow = slow.next;
            fast = fast.next;
            if(fast!=null){
                fast = fast.next;
            }
        }

        // 1 -> 2 -> 3 -> 4 ->5

        slowPrev.next = null;
        //slow -> is the head of 2 nd list

        //Reverse the 2nd list

        ListNode reverseHead = reverse(slow); // Step 2 => Reverse the 2nd list

        ListNode mergedHead = new ListNode(-1);
        ListNode it = mergedHead;
        ListNode it1 = head; // First List head
        ListNode it2 = reverseHead; // Second List head

        while(it1!=null && it2!=null){ // Merge the 2 lists alternatively
            ListNode for1 = it1.next;
            it.next = it1;
            it1 = for1;

            it = it.next;

            ListNode for2 = it2.next;
            it.next = it2;
            it2 = for2;

            it = it.next;
        }

        head = mergedHead.next; // Set back to head
        return ;

    }

    private ListNode reverse(ListNode slow){
        ListNode prev =null;
        ListNode curr = slow;
        while(curr!=null){
            ListNode forward = curr.next;
            curr.next = prev;
            prev = curr;
            curr = forward;
        }
        return prev;
    }

}

File: December2021/C++/Reorder_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        if(!head->next || !head->next->next) return ;
        ListNode* prev=NULL,*slow=head,*fast=head;
        while(fast && fast->next)
        {
            prev=slow;
            slow=slow->next;
            fast=fast->next->next;
        }
        
        prev->next=NULL;
        slow=reverse(slow);
        
        merge(head,slow);
        
    }
    ListNode* reverse(ListNode* node)
    {
        ListNode* curr=node;
        ListNode* prev=NULL,*nextnode=node->next;
        while(curr)
        {
            curr->next=prev;
            prev=curr;
            curr=nextnode;
            nextnode=nextnode?nextnode->next:nextnode;
        }
        return prev;
        
    }
    
    void merge(ListNode* l1,ListNode* l2)
    {
        ListNode* n1,*n2;
        while(l1 && l2)
        {
            n1=l1->next;n2=l2->next;
            l1->next=l2;
            if(!n1) return ;
            l2->next=n1;
            l1=n1;
            l2=n2;
        }
    }
};